dm/pkg/utils/db.go

```diff
@@ -391,49 +391,55 @@ func GetServerUnixTS(ctx context.Context, db *sql.DB) (int64, error) {
 	return ts, err
 }
 
-// GetCharsetAndDefaultCollation gets charset and default collation map.
-func GetCharsetAndDefaultCollation(ctx context.Context, db *sql.DB) (map[string]string, error) {
+// GetCharsetAndCollationInfo gets charset and collation info.
+func GetCharsetAndCollationInfo(ctx context.Context, db *sql.DB) (map[string]string, map[int]string, error) {
 	charsetAndDefaultCollation := make(map[string]string)
+	idAndCollationMap := make(map[int]string)
+
 	conn, err := db.Conn(ctx)
 	if err != nil {
-		return nil, terror.DBErrorAdapt(err, terror.ErrDBDriverError)
+		return nil, nil, terror.DBErrorAdapt(err, terror.ErrDBDriverError)
 	}
 	defer conn.Close()
 
 	// Show an example.
 	/*
-		mysql> SHOW CHARACTER SET;
-		+----------+---------------------------------+---------------------+--------+
-		| Charset  | Description                     | Default collation   | Maxlen |
-		+----------+---------------------------------+---------------------+--------+
-		| armscii8 | ARMSCII-8 Armenian              | armscii8_general_ci |      1 |
-		| ascii    | US ASCII                        | ascii_general_ci    |      1 |
-		| big5     | Big5 Traditional Chinese        | big5_chinese_ci     |      2 |
-		| binary   | Binary pseudo charset           | binary              |      1 |
-		| cp1250   | Windows Central European        | cp1250_general_ci   |      1 |
-		| cp1251   | Windows Cyrillic                | cp1251_general_ci   |      1 |
-		+----------+---------------------------------+---------------------+--------+
+		mysql> SELECT COLLATION_NAME,CHARACTER_SET_NAME,ID,IS_DEFAULT from INFORMATION_SCHEMA.COLLATIONS;
+		+----------------------------+--------------------+-----+------------+
+		| COLLATION_NAME             | CHARACTER_SET_NAME | ID  | IS_DEFAULT |
+		+----------------------------+--------------------+-----+------------+
+		| armscii8_general_ci        | armscii8           |  32 | Yes        |
+		| armscii8_bin               | armscii8           |  64 |            |
+		| ascii_general_ci           | ascii              |  11 | Yes        |
+		| ascii_bin                  | ascii              |  65 |            |
+		| big5_chinese_ci            | big5               |   1 | Yes        |
+		| big5_bin                   | big5               |  84 |            |
+		| binary                     | binary             |  63 | Yes        |
+		+----------------------------+--------------------+-----+------------+
 	*/
 
-	rows, err := conn.QueryContext(ctx, "SHOW CHARACTER SET")
+	rows, err := conn.QueryContext(ctx, "SELECT COLLATION_NAME,CHARACTER_SET_NAME,ID,IS_DEFAULT from INFORMATION_SCHEMA.COLLATIONS")
```

Comment:
cc @okJiang we have extra requirements for privilege precheck of incremental phase 😂

Reply:
How about we checking all INFORMATION_SCHEMA's SELECT privilege?

Reply:
Does this query from `information_schema.collations` causes dead lock if there are other running tasks at dump phase? /cc @lichunzhu PTAL

Reply:
It won't cause a dead lock. But sometimes a query to `information_schema` could be unacceptably slow.